src/main/kotlin/com/atlassian/performance/tools/awsinfrastructure/api/network/access/LocalPublicIpv4Provider.kt

```diff
@@ -0,0 +1,61 @@
+package com.atlassian.performance.tools.awsinfrastructure.api.network.access
+
+import com.atlassian.performance.tools.awsinfrastructure.Ipv4Validator
+import com.atlassian.performance.tools.awsinfrastructure.VerboseUrlReader
+import java.net.URL
+import java.util.function.Supplier
+
+class LocalPublicIpv4Provider private constructor(
+    private val servicesToQuery: List<URL>,
+    private val urlReader: (serviceUrl: URL) -> String?,
+    private val ipValidator: (ip: String) -> Boolean
+) : () -> String, Supplier<String> {
+    object Defaults {
+        val servicesToQuery: List<URL> = listOf(
+            URL("https://checkip.amazonaws.com")
+        )
+        val urlReader: (serviceUrl: URL) -> String? = VerboseUrlReader()
+        val ipValidator: (ip: String) -> Boolean = Ipv4Validator()
+    }
+
+    constructor() : this(
+        servicesToQuery = Defaults.servicesToQuery,
+        urlReader = Defaults.urlReader,
+        ipValidator = Defaults.ipValidator
+    )
+
+    override fun invoke() = get()
+
+    override fun get() = servicesToQuery
+        .asSequence()
+        .mapNotNull { serviceUrl -> urlReader(serviceUrl)?.trim() }
+        .filter { it.isNotEmpty() }
+        .filter { ipValidator(it) }
+        .toSet()
+        .let { set ->
+            if (set.isEmpty()) {
+                throw IllegalStateException("Queried services didn't report any valid public IP for this machine")
+            }
+            if (set.size != 1) {
+                throw IllegalStateException("Queried services reported different public IPs for this machine: ${set.toList()}")
```

Comment:
different -> multiple? Also, I'm not sure if this message is actionable for the user. What are they supposed to do in such a case?

Reply:
The logic was supposed to give an opportunity to the API consumers to define many services in case one wouldn't be stable enough. In such case there can be a situation where those won't return the same IP. This situation should be treated seriously as either misconfiguration (e.g. one of the services used is broken or going though different [NIC](https://en.wikipedia.org/wiki/Network_interface_controller)) or something malicious happening (e.g. someone is doing [MITM attack](https://en.wikipedia.org/wiki/Man-in-the-middle_attack)). I can note the possibilities in the error message.

Alterntively I could make it less strict here and always take first response no matter if there were 2 or more different ones. The concern I have is that it could hide potential problem, which could have not obvious symptoms otherwise. The lib is opensource, so if the API consumer is experienced enough they should be able to easily find the error message in the code and reason based on it. Otherwise the consumer should come to us for support and it should also not be to hard to deduce what can be done do diagnose the problem further.

BTW the default service i added is one that is not documented by AWS. I'm ready to pick different one and possibly have more then one in the list for stability. Not sure which ones though.

Reply:
OK, let's keep it as is.